### PR TITLE
fix: Missing pi rows for event programs [DHIS2-12816] (2.39)

### PR DESCRIPTION
**Pull request by ameenhere**
There have been cases reported where a required row in the programinstance table for corresponding EVENT Programs is missing. An example in the SL demo is for the event program "`XX MAL RDT - Case Registration`"

This PR adds a java based flyway migration that tries to fill in the missing row.

### CODE REVIEW

File: dhis-2/dhis-support/dhis-support-db-migration/src/main/java/org/hisp/dhis/db/migration/v39/V2_39_17__Add_missing_programinstance_rows_for_programs_without_registration.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+/**
+ * Populates the missing programinstance row that has to be present exactly once
+ * for every program without registration.
+ *
+ * @author Ameen Mohamed
+ */
+@Slf4j
+public class V2_39_17__Add_missing_programinstance_rows_for_programs_without_registration extends BaseJavaMigration
+{
+    private static final String FETCH_PROGRAMS_MISSING_PI_SQL = " select pi.programinstanceid,p.programid from program p left join programinstance pi on p.programid=pi.programid where p.type='WITHOUT_REGISTRATION' and pi.programinstanceid is null";
```
Comment (larshelge) [inline dhis-2/dhis-support/dhis-support-db-migration/src/main/java/org/hisp/dhis/db/migration/v39/V2_39_17__Add_missing_programinstance_rows_for_programs_without_registration.java:53]: Not critical, but perhaps a subquery would be more readable here and make the intent clearer, i.e.

> select p.programid from program p where not exists (select pi.programinstanceid from programinstance pi where pi.programid = p.programid);

**Review by teleivo — COMMENTED** [review]
👍🏻 on looking into what @zubaira said

| Bulk insert/import can be used instead of inserting in the loop.

Even if there are not many instances of such missing rows. Its a good habit to make migrations as quick as we can in case there are many rows that need to be inserted/changed.

Do you also know why that happens? If so did you/do you plan to fix that as well, which would be equally important 😋

File: dhis-2/dhis-support/dhis-support-db-migration/src/main/java/org/hisp/dhis/db/migration/v39/V2_39_17__Add_missing_programinstance_rows_for_programs_without_registration.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        {
+
+            while ( rs.next() )
+            {
+                programsWithMissingDefaultProgramInstance.add( rs.getLong( "programid" ) );
+            }
+        }
+        catch ( SQLException e )
+        {
+            log.error( "Flyway java migration error:", e );
```
[the remark this answers is not in the document]
Comment (ameenhere) [reply dhis-2/dhis-support/dhis-support-db-migration/src/main/java/org/hisp/dhis/db/migration/v39/V2_39_17__Add_missing_programinstance_rows_for_programs_without_registration.java]: Agreed. Removed the logging

**Comment by ameenhere** [comment]
> Do you also know why that happens? If so did you/do you plan to fix that as well, which would be equally important

Pretty sure this had to be some sort of direct modification in the database done my sys admins. In our play db, might have just been a lingering corrupted data from long back.

**Review by enricocolasante — APPROVED** [review]

**Pull request merged**